Add ServerDrivesInfo() admin API

nitisht opened the pull request:
## Description
This is a part of implementation of mc admin health command.
ServerDrivePerfInfo() exposes local and remote server drive mount path
and write speed information.

## Motivation and Context
Part of https://github.com/minio/mc/issues/2606

## Regression
No

## How Has This Been Tested?
Using the madmin package

```go
	// Initialize minio client object.
	mdmClnt, err := madmin.New("127.0.0.1:9000", "minio", "minio123", ssl)
	if err != nil {
		fmt.Println(err)
		return
	}
	si, err := mdmClnt.ServerDrivesPerfInfo()
	if err != nil {
		log.Fatalln(err)
	}
	b, err := json.Marshal(si)
	fmt.Printf("%s\n", string(b))
```

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -284,6 +284,73 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
+// ServerDrivePerfInfo holds informantion about address and write speed of all drives in a single server node
+type ServerDrivePerfInfo struct {
+	Addr       string            `json:"addr"`
+	DrivesPerf []DriveWriteSpeed `json:"drivesPerf"`
```
Perf should contain Write and Read speed as we might do Reads as well in future @nitisht

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -284,6 +284,73 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
+
+// DriveWriteSpeed holds informantion about write speed of a single drive
+type DriveWriteSpeed struct {
+	Path       string  `json:"path"`
```
This should be called Disk instead of Path, perhaps also its endpoint representation? such that we know which server the disk belongs to.

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -284,6 +284,73 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
+			// Initialize server info at index
+			reply[idx] = ServerDrivePerfInfo{Addr: peer.addr}
+
+			drivesPerf, err := peer.cmdRunner.ServerDrivePerfInfo()
```
Use the notification.go for this we are going to deprecate the user of peer.cmdRunner @nitisht

harshavardhana commented on cmd/local-admin-client.go:
```diff
@@ -110,3 +115,54 @@ func (lc localAdminClient) DownloadProfilingData() ([]byte, error) {
+		// Only use local endpoints here
+		if endpoint.IsLocal {
+			if _, err := os.Stat(endpoint.Path); err != nil {
+				// Since this drive/endpoint is not available, continue
```
Comment can be better

harshavardhana commented on cmd/local-admin-client.go:
```diff
@@ -110,3 +115,54 @@ func (lc localAdminClient) DownloadProfilingData() ([]byte, error) {
+				fsTmpObjPath := pathJoin(endpoint.Path, minioMetaTmpBucket, tempObj)
+
+				start := time.Now()
+				_, err := fsCreateFile(context.Background(), fsTmpObjPath, data, nil, data.Size())
```
Do not use fsCreateFile, it has overahead create custom functions for this which are well suited for performance needs.

harshavardhana commented on cmd/local-admin-client.go:
```diff
@@ -110,3 +115,54 @@ func (lc localAdminClient) DownloadProfilingData() ([]byte, error) {
+				elapsed := time.Since(start)
+
+				if err != nil {
+					return drivePerf, errServerNotInitialized
```
Incorrect error to be returned here @nitisht - if disk create failed we should return appropriate error something from posix-errors.go

harshavardhana commented on cmd/local-admin-client.go:
```diff
@@ -110,3 +115,54 @@ func (lc localAdminClient) DownloadProfilingData() ([]byte, error) {
+				}
+				totalElapsed += elapsed
+				// remove the temp file created
+				fsRemoveFile(context.Background(), fsTmpObjPath)
```
Same as fsCreateFile do not use fs-v1-helpers.go create your own wrappers.

harshavardhana commented on cmd/local-admin-client.go:
```diff
@@ -110,3 +115,54 @@ func (lc localAdminClient) DownloadProfilingData() ([]byte, error) {
+			// Average time spent = total time elapsed / total times uploaded
+			averageTime := totalElapsed.Seconds() / float64(writeCount)
+			// Write perf = fileSize (in MB) / average time spent (in seconds)
+			driveInfo.WriteSpeed = readSizeV1 / (humanize.MiByte * averageTime)
```
Ideally, I think you should simply extend `pkg/disk` to have this feature. Such that you can have a call like `GetPerformanceInfo()` which returns Performance information about the disk, includes write and read speeds.

nitisht committed:
Add ServerDrivesPerfInfo() admin API

This is part of implementation for mc admin health command. The
ServerDrivesPerfInfo() admin API returns read and write speed
information for all the drives (local and remote) in a given Minio
server deployment.

Part of minio/mc#2606

harshavardhana commented on cmd/notification.go:
```diff
@@ -512,6 +512,29 @@ func (sys *NotificationSys) Send(args eventArgs) []event.TargetIDErr {
+				di.Addr = addr.String()
+				di.Error = err.Error()
+			}
+			reply = append(reply, di)
```
This won't work inside a Go routine, this causes races. You should always use a preallocated slice upto the length of `peerClientMap`  and then increment `i++` after `Go func()`

nitisht replied:
done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +44,106 @@ type Info struct {
+	// Create a file 10 times in series, then find the average write time
+	for writeCount = 1; writeCount <= 10; writeCount++ {
+		rand.Read(buf)
+		tempObj := mustGetUUID()
```
Don't use mustGetUUID() here, the fsPath passed here should already be pointing to a file which can be written to or read from, no need to have such duplicated code.

nitisht replied:
done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +44,106 @@ type Info struct {
+
+		start := time.Now()
+		n, err := writeFile(fsTempObjPath, buf)
+		elapsed := time.Since(start)
```
elasped can be calculated after success.

nitisht replied:
done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +44,106 @@ type Info struct {
+		elapsed := time.Since(start)
+
+		if err != nil {
+			removeFile(fsTempObjPath)
```
you can do a `defer removeFile(fsTempObjPath)`

nitisht replied:
done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +44,106 @@ type Info struct {
+	buf := make([]byte, fileSize)
+	rand.Read(buf)
+	// Create a file to read
+	n, err := writeFile(fsTempObjPath, buf)
```
Read shouldn't recreate a new file IMO, read and write should happen in the same function. This avoids you writing the file again.

nitisht replied:
makes sense, done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+		n, err := writeFile(fsTempObjPath, buf)
+
+		if err != nil {
+			cleanup(fsPath)
```
Once deferred you don't need this again here.

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+			return 0, 0, err
+		}
+		if n != fileSize {
+			cleanup(fsPath)
```
Once deferred you don't need this again here.

harshavardhana commented on pkg/disk/helpers.go:
```diff
@@ -0,0 +1,56 @@
+	return n, nil
+}
+
+func cleanup(path string) error {
```
Why provide a wrapped function ?

nitisht replied:
removed

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+
+	// create the enclosing directory
+	err = os.Mkdir(fsPath, 0777)
+	if err != nil {
```
Do you want to do os.MkdirAll ??

nitisht replied:
yeah, that'd be better

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+
+	for count = 1; count <= 10; count++ {
+		rand.Read(buf)
+		fsTempObjPath := path.Join([]string{fsPath, strconv.Itoa(count)}...)
```
path.Join doesn't require []string

Do just this

```
path.Join(dir1, dir2, ...)
```

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+		return 0, 0, err
+	}
+
+	for count = 1; count <= 10; count++ {
```
Make this count configurable default 10 might need to be changed in future.

nitisht replied:
okay done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +46,87 @@ type Info struct {
+	}
+
+	for count = 1; count <= 10; count++ {
+		rand.Read(buf)
```
We can avoid doing rand.Read() since it is taxing CPU you can simply use bytes.Repeat perhaps or write just zeros for this.

nitisht replied:
use rand.Read() once for 1 KiB and then using bytes.Repeat

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +51,90 @@ type Info struct {
+// Calculate the write and read performance - write and read 10 tmp (1 MiB)
+// files and find the average time taken (Bytes / Sec)
+func performance(fsPath string) (write, read float64, err error) {
+	var count int
```
You can do ` 		defer os.RemoveAll(fsPath)` right here and never have to do os.RemoveAll before return elsewhere. Since these files need to be cleaned up anyways

nitisht replied:
done

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +51,90 @@ type Info struct {
+
+// Calculate the write and read performance - write and read 10 tmp (1 MiB)
+// files and find the average time taken (Bytes / Sec)
+func performance(fsPath string) (write, read float64, err error) {
```
maybe function name can be not so generic? `doPerformance` ? might be better?

nitisht replied:
changed to `doPerfMeasure`

harshavardhana commented on pkg/disk/disk.go:
```diff
@@ -32,3 +51,90 @@ type Info struct {
+	}
+
+	for count = 1; count <= perfCount; count++ {
+		buf := bytes.Repeat(randBuf, randParts)
```
since we use this buffer repeatedly without change we don't need to do this each and every time in the loop we can just do this once for the entire perfCount() IMO

harshavardhana replied:
Also perf count can be call `iterations` ?

nitisht replied:
> we can just do this once for the entire perfCount

ah right, done

harshavardhana commented on pkg/disk/helpers.go:
```diff
@@ -0,0 +1,52 @@
+/*
+ * Minio Cloud Storage, (C) 2015 Minio, Inc.
```
Copyright is 2018

nitisht replied:
done

harshavardhana commented on cmd/admin-handlers.go:
```diff
@@ -284,6 +285,48 @@ func (a adminAPIHandlers) ServerInfoHandler(w http.ResponseWriter, r *http.Reque
+	Perf  []disk.Performance `json:"perf"`
+}
+
+// DrivesPerfInfoHandler - GET /minio/admin/v1/health/drive
```
Perhaps we should make this a bit more extensible by not having a resource instead have a query param, also /v1/health is an odd endpoint for getting disk performance. We can retro fit it .. or have something like 

/v1/performance?type=drive
/v1/performance?type=network 

??

nitisht replied:
yeah that is good idea, I guess that way it is nicely extensible, changing to this approach

nitisht replied:
done

harshavardhana reviewed (changes requested):
Everything else LGTM, only question is the extensibility part.

harshavardhana commented on pkg/madmin/info-commands.go:
```diff
@@ -147,3 +150,51 @@ func (adm *AdminClient) ServerInfo() ([]ServerInfo, error) {
+type PerfType string
+
+// ServerDrivesPerfInfo - server's endpoints performance info based on input PerfType
+func (adm *AdminClient) ServerDrivesPerfInfo(perfType PerfType) ([]ServerDrivesPerfInfo, error) {
```
You don't need to take drive as perf type because this function is called ServerDrivesPerfInfo

nitisht replied:
done

harshavardhana reviewed (dismissed):
LGTM

harshavardhana commented on cmd/admin-router.go:
```diff
@@ -53,6 +53,9 @@ func registerAdminRouter(router *mux.Router, enableIAM bool) {
 	adminV1Router.Methods(http.MethodGet).Path("/info").HandlerFunc(httpTraceAll(adminAPI.ServerInfoHandler))
 
+	// Performance command - retrun performance details based on input type
+	adminV1Router.Methods(http.MethodGet).Path("/performance").HandlerFunc(httpTraceAll(adminAPI.PerfInfoHandler)).Queries("perfType", "{perfType:.*}")
```
This should be moved into XL only @nitisht as @kannappanr mentioned

nitisht replied:
> This should be moved into XL only

But we may want to return disk perf for FS deployments too. How about adding a check in `PerfInfoHandler`?

harshavardhana replied:
There is no value in running for FS mode IMO - this is generally to centrally run across many disks. FS mode one can simply run `dd`

nitisht replied:
okay, updated. @harshavardhana @kannappanr PTAL

harshavardhana reviewed (dismissed):
LGTM

kannappanr reviewed (approved):
LGTM and tested.
